cardano-api/src/Cardano/Api/IO.hs

```diff
@@ -51,10 +54,25 @@ import qualified Data.ByteString.Lazy as LBSC
 import           Data.String (IsString)
 import           Data.Text (Text)
 import qualified Data.Text.IO as Text
-import           GHC.Generics (Generic)
 import qualified System.IO as IO
 import           System.IO (Handle)
 
+import           Cardano.Api.Error (FileError (..))
+
+data FileDirection
+  = In
+  -- ^ Indicate the file is to be used for reading.
+  | Out
+  -- ^ Indicate the file is to be used for writing.
+  | InOut
+  -- ^ Indicate the file is to be used for both reading and writing.
+
+-- | A file path with additional type information to indicate what the file is meant to
+-- contain and whether it is to be used for reading or writing.
+newtype File content (direction :: FileDirection) = File
```

Comment:
`direction` could also be represented as a type-level list `[FileDirection]` . This way functions requiring file for writing for example could have a following signature:
```haskell
writeByteStringOutput :: ()
  => MonadIO m
  => HasFileDirection Out directions
  => Maybe (File content directions)
  -> ByteString
  -> m (Either (FileError ()) ())
```
this way conversion of `InOut` to `In`  or `Out` would be not necessary and `onlyOut` functions wouldn't be needed.

Reply:
I like this idea, but it can be a separate PR.